File: src/core/sources.js
'use strict';

// 采集源健康检查：目录是否存在、会话文件数量、空态文案。
// 与聚合解耦，供 Store / CLI / UI 共用。

const fs = require('fs');
const { rootOf } = require('../shared/paths');

const SOURCE_META = {
  claude: {
    label: 'Claude Code',
    hint: '~/.claude/projects/*/*.jsonl',
    how: '使用 Claude Code 产生会话后会出现用量',
  },
  codex: {
    label: 'Codex',
    hint: '~/.codex/sessions/**/*.jsonl',
    how: '使用 Codex 产生会话后会出现用量',
  },
  grok: {
    label: 'Grok Build',
    hint: '~/.grok/sessions/**/updates.jsonl',
    how: '使用 Grok Build 产生会话后会出现用量',
  },
  pi: {
    label: 'Pi',
    hint: '~/.pi/agent/sessions/**/*.jsonl',
    how: '使用 Pi 产生会话后会出现用量',
  },
};

// 探测单个源：missing | empty | ok | error
function probeOne(name, collector) {
  const meta = SOURCE_META[name] || { label: name, hint: '', how: '' };
  const root = rootOf(name);
  const base = {
    key: name,
    label: meta.label,
    root,
    hint: meta.hint,
    how: meta.how,
    exists: false,
    fileCount: 0,
    status: 'missing',
    message: '',
  };

  try {
    if (!fs.existsSync(root)) {
      base.status = 'missing';
      base.message = `目录不存在：${root}`;
      return base;
    }
    base.exists = true;
    let files = [];
    try {
      files = collector.listSourceFiles() || [];
    } catch (err) {
      base.status = 'error';
      base.message = String(err && err.message ? err.message : err);
      return base;
    }
    base.fileCount = files.length;
    if (files.length === 0) {
      base.status = 'empty';
      base.message = `已找到目录，但尚无会话文件（${meta.hint}）`;
      return base;
    }
    base.status = 'ok';
    base.message = `${files.length} 个会话文件`;
    return base;
  } catch (err) {
    base.status = 'error';
    base.message = String(err && err.message ? err.message : err);
    return base;
  }
}

// collectors: { claude, codex, grok, pi }
function probeSources(collectors) {
  const tools = {};
  let missing = 0;
  let empty = 0;
  let errors = 0;
  let totalFiles = 0;

  for (const [name, collector] of Object.entries(collectors)) {
    const info = probeOne(name, collector);
    tools[name] = info;
    totalFiles += info.fileCount || 0;
    if (info.status === 'missing') missing++;
    else if (info.status === 'empty') empty++;
    else if (info.status === 'error') errors++;
  }

  const names = Object.keys(tools);
  const allQuiet = totalFiles === 0; // 四源都没有可解析文件
  const lines = [];
  for (const name of names) {
    const t = tools[name];
    if (t.status !== 'ok') {
      lines.push(`${t.label}：${t.message}`);
    }
  }

  let banner = '';
  if (allQuiet) {
    banner =
      '暂无用量数据。请确认已安装并使用过 Claude Code / Codex / Pi / Grok Build；会话目录见各卡片提示。';
  } else if (missing > 0 || errors > 0) {
    banner = `部分数据源不可用（缺失 ${missing} · 错误 ${errors}）。详见状态栏或 CLI。`;
  }

  return {
    tools,
    totalFiles,
    missing,
    empty,
    errors,
    allQuiet,
    banner,
    issues: lines,
  };
}

// 生成状态栏短文案
function sourcesStatusLine(sources) {
  if (!sources) return '';
  if (sources.allQuiet) return '无会话数据';
  const parts = [];
  for (const t of Object.values(sources.tools || {})) {
    if (t.status === 'ok') continue;
    if (t.status === 'missing') parts.push(`${t.label}目录缺失`);
    else if (t.status === 'empty') parts.push(`${t.label}无会话`);
    else if (t.status === 'error') parts.push(`${t.label}异常`);
  }
  return parts.join(' · ');
}

module.exports = {
  SOURCE_META,
  probeOne,
  probeSources,
  sourcesStatusLine,
};

File: src-tauri/src/core/sources.rs
use super::collectors::{home_dir, list_jsonl_files, list_jsonl_files_recursive};
use super::types::{SourceInfo, SourcesResult};
use std::collections::HashMap;
use std::path::PathBuf;

struct ProbeDef {
    key: &'static str,
    label: &'static str,
    root: fn() -> PathBuf,
    hint: &'static str,
    how: &'static str,
    recursive: bool,
    filename_filter: Option<&'static str>,
}

const PROBES: &[ProbeDef] = &[
    ProbeDef {
        key: "claude",
        label: "Claude Code",
        root: || home_dir().join(".claude").join("projects"),
        hint: "~/.claude/projects/*/*.jsonl",
        how: "使用 Claude Code 产生会话后会出现用量",
        recursive: false,
        filename_filter: None,
    },
    ProbeDef {
        key: "codex",
        label: "Codex",
        root: || home_dir().join(".codex").join("sessions"),
        hint: "~/.codex/sessions/**/*.jsonl",
        how: "使用 Codex 产生会话后会出现用量",
        recursive: true,
        filename_filter: None,
    },
    ProbeDef {
        key: "pi",
        label: "Pi",
        root: || home_dir().join(".pi").join("agent").join("sessions"),
        hint: "~/.pi/agent/sessions/**/*.jsonl",
        how: "使用 Pi 产生会话后会出现用量",
        recursive: true,
        filename_filter: None,
    },
    ProbeDef {
        key: "grok",
        label: "Grok Build",
        root: || home_dir().join(".grok").join("sessions"),
        hint: "~/.grok/sessions/**/updates.jsonl",
        how: "使用 Grok Build 产生会话后会出现用量",
        recursive: false,
        filename_filter: Some("updates.jsonl"),
    },
];

pub fn probe_sources() -> SourcesResult {
    let mut tools = HashMap::new();
    let mut total_files = 0u64;
    let mut missing = 0u64;
    let mut empty = 0u64;
    let mut errors = 0u64;
    let mut issues = Vec::new();
    let mut all_quiet = true;

    for probe in PROBES {
        let root = (probe.root)();
        let exists = root.exists();

        let file_count = if exists {
            if probe.recursive {
                list_jsonl_files_recursive(&root).len() as u64
            } else {
                list_jsonl_files(&root).len() as u64
            }
        } else {
            0
        };

        let (status, message) = if !exists {
            missing += 1;
            all_quiet = false;
            ("missing".to_string(), format!("目录 {} 未创建", root.display()))
        } else if file_count == 0 {
            empty += 1;
            ("empty".to_string(), "目录已创建但尚未有会话文件".to_string())
        } else {
            ("ok".to_string(), format!("{} 个会话文件", file_count))
        };

        tools.insert(
            probe.key.to_string(),
            SourceInfo {
                key: probe.key.to_string(),
                label: probe.label.to_string(),
                root: root.display().to_string(),
                hint: probe.hint.to_string(),
                how: probe.how.to_string(),
                exists,
                file_count,
                status,
                message,
            },
        );

        total_files += file_count;
    }

    let banner = if all_quiet && total_files == 0 {
        "暂无数据：所有数据源均为空状态。请先使用 AI 编码工具产生一些会话记录。".to_string()
    } else {
        String::new()
    };

    SourcesResult {
        tools,
        total_files,
        missing,
        empty,
        errors,
        all_quiet,
        banner,
        issues,
    }
}
